api/task_manager.cc

```diff
@@ -0,0 +1,202 @@
+#include <seastar/core/coroutine.hh>
+
+#include "task_manager.hh"
+#include "api/api-doc/task_manager.json.hh"
+#include "db/system_keyspace.hh"
+#include "column_family.hh"
+#include "unimplemented.hh"
+#include "storage_service.hh"
+
+#include <utility>
+
+namespace api {
+
+namespace tm = httpd::task_manager_json;
+using namespace json;
+
+inline bool filter_tasks(seastar::shared_ptr<task_manager::task> task, std::unordered_map<sstring, sstring>& query_params) {
+    return (!query_params.contains("keyspace") || query_params["keyspace"] == task->get_status().keyspace) &&
+        (!query_params.contains("table") || query_params["table"] == task->get_status().table);
+}
+
+struct status {
+    task_manager::task_status task_status;
+    task_manager::running_tasks number;
+    task_manager::module module;
+    unsigned shard;
+};
+
+tm::task_status make_status(status status) {
+    auto start_time = gc_clock::to_time_t(status.task_status.start_time);
+    auto end_time = gc_clock::to_time_t(status.task_status.end_time);
+    ::tm st, et;
+    ::gmtime_r(&end_time, &et);
+    ::gmtime_r(&start_time, &st);
+
+    tm::task_status res{};
+    res.id = status.task_status.id.to_sstring();
+    res.type = status.task_status.type;
+    res.state = status.task_status.state;
+    res.is_abortable = status.task_status.is_abortable;
+    res.start_time = st;
+    res.end_time = et;
+    res.error = status.task_status.error;
+    res.parent_id = status.task_status.parent_id.to_sstring();
+    res.number = status.number.get_number(status.module);
```

Comment:
Why do we need to call get_number?
I presume the task number can be assigned when it is created and be simply returned in the status.
Modules that don't enumerate their tasks (i.e. compaction in contrast to repair) should just initialize that to 0 (you can define the default value when defining `task_manager::task_status` (or should be a sub-class of task, i.e. `task_manager::task::status`?)